clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -135,6 +146,7 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
     public void transitionToFenced() {
         resetEpoch();
         transitionTo(MemberState.FENCED);
+        assignmentReconciler.startLost();
```

Comment:
the sequence here is not right I would say. Transitioning to FENCED currently means that the next poll will send out a HB with epoch 0, which should only be done when the `onPartitionsLost` callback completes. Wondering if an intermediate state like RELEASING_ASSIGNMENT would do. Similar to RECONCILING, but different because we do need to continue HB while reconciling, but not while releasing, because when fenced the member is not part of the group anymore until it rejoins. Both, reconciling and releasing would complete when the callbacks complete, but reconciling would transition to stable, vs. releasing that would transition to fenced (triggering the re-join). Makes sense?

Reply:
I think this is actually fine, because the heartbeat won't be sent in the FENCED state. I see that completeLost() is invoked after the callback completes - there I think we can ask the member to rejoin by invoking tryJoin() (which is in my PR).

Technically - The member should try to rejoin after being fenced, but there's no guarantee to it. In this case - we rely on callback completion signal to ask the member to rejoin.  This is what i meant.

Reply:
Uhm why do you say that "_the heartbeat won't be sent in the FENCED state_"? , that's not true in the current implementation. With the logic we have so far, the HB manager poll does generate a HB request on the interval in the FENCED state.

Reply:
sorry, you are right, i kept forgetting about the changes - after reading the PR, due to the fact that we have to wait for the rebalance to complete to rejoin, maybe it is best to first move to the next state, i.e. UNJOINED (or something else is fine), and then send the heartbeat.